odoo/orm/registry.py

```diff
@@ -554,34 +554,46 @@ def post_init(self, func, *args, **kwargs):
-                # Module A may try to apply a constraint and fail but another module B inheriting
-                # from Module A may try to reapply the same constraint and succeed, however the
-                # constraint would already be in the _constraint_queue and would be executed again
-                # at the end of the registry cycle, this would fail (already-existing constraint)
-                # and generate an error, therefore a constraint should only be applied if it's
-                # not already marked as "to be applied".
-                func(*args, **kwargs)
-        except Exception as e:
-            if self._is_install:
-                _schema.error(*e.args)
-            else:
-                _schema.info(*e.args)
-                self._constraint_queue.append((func, args, kwargs))
+            func_data[0](*func_data[1:])
+        except sql.ConstraintError as e:
+            _schema.info(e.args[0])
+            # Try to apply the constraint once the registry is fully loaded. We keep multiple
+            # constraints under same key because we still want to attempt to add a constraint
+            # even if some override failed. Do not add duplicated entries, if an entry is double
+            # move it to the end instead. Below is a poor-man move-to-end implementation using a
+            # standard dict.
+            self._constraint_stacks[key].pop(func_data, None)
+            self._constraint_stacks[key][func_data] = None
+        else:
+            # If a module successfully applies a constraint, we remove all queued constraints with
+            # same name ignoring the definition. In effect, if module A defines a constraint that is
+            # overridden by module B (loaded after) with a different definition, whenever we load A,
+            # the original constraint may fail to be applied, hence it is queued. The issue is that
+            # later when we post process the queue the constraint still fails to be added. The
+            # solution is to clean the previously enqeueud constraints that were just overriden.
+            same = self._constraint_stacks.pop(key, None)
+            if same:
+                _schema.info("Skipping already applied constraints: %s", list(same))
 
     def finalize_constraints(self):
         """ Call the delayed functions from above. """
-        while self._constraint_queue:
-            func, args, kwargs = self._constraint_queue.popleft()
-            try:
-                func(*args, **kwargs)
-            except Exception as e:
-                # warn only, this is not a deployment showstopper, and
-                # can sometimes be a transient error
-                _schema.warning(*e.args)
+        for key, funcs_data in reversed(self._constraint_stacks.items()):
+            errors = []
+            for func_data in reversed(funcs_data):
+                try:
+                    func_data[0](*func_data[1:])
```

Comment:
Call the function within a savepoint to be able to resume it.

Reply:
IMO, the handling of the savepoint should be done in `sql.add_index` since `sql.add_constraint` also handles it, as well as `sql.set_not_null`. I renamed the exception to something more general like `TableError` (could be either due to a constraint error, or an index error, suggestions for a better naming are welcome), same for the stacks of changes (renamed to _table_stacks)

The new patch is rebased in latest master.

wdyt?

Reply:
I will add the savepoint in `sql.add_index` as a fix.

Reply:
> Note that adding foreign keys does not use a savepoint either.

IMO that's fine. Not being able to add a FK is really a fundamental problem that should block everything. But indexes and constraints are points of extension in Odoo. Any inheriting model can alter them somehow, and thus depending on the loading order we expect some of them to fail... but not all. But even if they fail those are not as fundamental as FKs. My 2 cents.